orttraining/orttraining/python/training/ortmodule/_utils.py

```diff
@@ -56,21 +58,43 @@ def _ortvalue_from_torch_tensor(torch_tensor):
     return C.OrtValue.from_dlpack(to_dlpack(torch_tensor), is_bool_tensor)
 
 
-def _torch_tensor_from_dl_pack(dlpack, ortvalue):
-    torch_tensor = from_dlpack(dlpack)
-    return torch_tensor.to(torch.bool) if ortvalue.data_type() == 'tensor(bool)' else torch_tensor
+def _ortvalues_to_torch_tensor(ortvalues, device):
+    if len(ortvalues) == 0:
+        return tuple()
+    if 'ort' == device.type:
+        if not hasattr(C, 'to_aten_ort_device_tensor'):
+            raise AttributeError("onnxruntime is missing to_aten_ort_device_tensor needed to support device == 'ort'.")
+        return tuple(C.to_aten_ort_device_tensor(ov) for ov in ortvalues)
+    if hasattr(ortvalues, 'to_dlpacks'):
+        res = ortvalues.to_dlpacks(_from_dlpack)
+        has_bool_tensor = ortvalues.has_bool_tensor()
```

Comment:
Maybe we can change has_bool_tensor() to bool_tensor_indices() to return all the indices of bool tensors in the list? So below IF is len(indices) > 0, and we can use this indices directly instead of another FOR loop to check the flag again.